src/index.js

```diff
@@ -170,224 +172,232 @@ class UglifyJsPlugin {
+              // Handling comment extraction
+              let commentsFile = false;
 
-            if (this.options.extractComments) {
-              commentsFile =
-                this.options.extractComments.filename || `${file}.LICENSE`;
+              if (this.options.extractComments) {
+                commentsFile =
+                  this.options.extractComments.filename || `${file}.LICENSE`;
 
-              if (typeof commentsFile === 'function') {
-                commentsFile = commentsFile(file);
+                if (typeof commentsFile === 'function') {
+                  commentsFile = commentsFile(file);
+                }
               }
-            }
 
-            const task = {
-              file,
-              input,
-              inputSourceMap,
-              commentsFile,
-              extractComments: this.options.extractComments,
-              uglifyOptions: this.options.uglifyOptions,
-              minify: this.options.minify,
-            };
-
-            if (this.options.cache) {
-              const defaultCacheKeys = {
-                // eslint-disable-next-line global-require
-                'uglify-js': uglifyJsPackageJson.version,
-                // eslint-disable-next-line global-require
-                'uglifyjs-webpack-plugin': require('../package.json').version,
-                'uglifyjs-webpack-plugin-options': this.options,
-                hash: crypto
-                  .createHash('md4')
-                  .update(input)
-                  .digest('hex'),
+              const task = {
+                file,
+                input,
+                inputSourceMap,
+                commentsFile,
+                extractComments: this.options.extractComments,
+                uglifyOptions: this.options.uglifyOptions,
+                minify: this.options.minify,
+                cacheKey,
               };
 
-              task.cacheKeys = this.options.cacheKeys(defaultCacheKeys, file);
+              tasks.push(task);
+            } catch (error) {
+              compilation.errors.push(
+                UglifyJsPlugin.buildError(
+                  error,
+                  file,
+                  UglifyJsPlugin.buildSourceMap(inputSourceMap),
+                  new RequestShortener(compiler.context)
+                )
+              );
             }
+          })
+      );
 
-            tasks.push(task);
-          } catch (error) {
-            compilation.errors.push(
-              UglifyJsPlugin.buildError(
-                error,
-                file,
-                UglifyJsPlugin.buildSourceMap(inputSourceMap),
-                new RequestShortener(compiler.context)
-              )
-            );
-          }
-        });
-
-      taskRunner.run(tasks, (tasksError, results) => {
-        if (tasksError) {
-          compilation.errors.push(tasksError);
+      promise.then(() => {
+        taskRunner.run(tasks, (tasksError, results) => {
```

Comment:
The callback to `taskRunner` is unchanged, I've only wrapped it in a `.then()` now that the computation of `tasks` is async.